cmd/prepare-storage.go

```diff
@@ -120,46 +120,65 @@ var errXLV3ThisEmpty = fmt.Errorf("XL format version 3 has This field empty")
 // connect to list of endpoints and load all XL disk formats, validate the formats are correct
 // and are in quorum, if no formats are found attempt to initialize all of them for the first
 // time. additionally make sure to close all the disks used in this attempt.
-func connectLoadInitFormats(firstDisk bool, endpoints EndpointList, setCount, drivesPerSet int) (*formatXLV3, error) {
+func connectLoadInitFormats(firstDisk bool, endpoints EndpointList, setCount, drivesPerSet int) (*formatXLV3, *sysInfoConfig, error) {
 	storageDisks, err := initStorageDisks(endpoints)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	defer closeStorageDisks(storageDisks)
 
 	// Attempt to load all `format.json` from all disks.
 	formatConfigs, sErrs := loadFormatXLAll(storageDisks)
 
+	// Attempt to load `sysinfo.json` from all disks.
+	sysInfoConfigs, sysInfoErrs := loadSysInfoXLAll(storageDisks)
+
 	// Pre-emptively check if one of the formatted disks
 	// is invalid. This function returns success for the
 	// most part unless one of the formats is not consistent
 	// with expected XL format. For example if a user is
 	// trying to pool FS backend into an XL set.
 	if err = checkFormatXLValues(formatConfigs); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	for i, sErr := range sErrs {
 		if _, ok := formatCriticalErrors[sErr]; ok {
-			return nil, fmt.Errorf("Disk %s: %s", endpoints[i], sErr)
+			return nil, nil, fmt.Errorf("Disk %s: %s", endpoints[i], sErr)
+		}
+	}
+	for i, sErr := range sysInfoErrs {
+		if _, ok := formatCriticalErrors[sErr]; ok {
+			return nil, nil, fmt.Errorf("Disk %s: %s", endpoints[i], sErr)
 		}
 	}
+	var sysinfo *sysInfoConfig
+
+	var format *formatXLV3
 
 	// All disks report unformatted we should initialized everyone.
 	if shouldInitXLDisks(sErrs) && firstDisk {
-		return initFormatXL(context.Background(), storageDisks, setCount, drivesPerSet)
+		format, err = initFormatXL(context.Background(), storageDisks, setCount, drivesPerSet)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if shouldCreateSysInfoConfig(sysInfoErrs) && firstDisk {
```

Comment:
It is probably better to include this block in the previous if and return (as it was before the change)

Reply:
The idea is to use it for upgrade scenario also. I still have to fix it properly, as during an upgrade it is still not creating a sysinfo.json file.